pym/bob/webdav.py

```diff
@@ -121,55 +109,66 @@ def upload(self, path, buf, overwrite):
-        if response.status == 412:
-            # precondition failed -> lost race with other upload
-            raise HttpAlreadyExistsError()
-        elif response.status not in [200, 201, 204]:
-            raise HttpUploadError("PUT {} {}".format(response.status, response.reason))
+
+        req = urllib.request.Request (self._getURL(path),
+                                      data=buf, headers=headers, method="PUT")
+        try:
+            with urllib.request.urlopen (req, context=self.__context) as resp:
+                if resp.status not in [200, 201, 204]:
+                    raise HttpUploadError("PUT {} {}".format(resp.status, resp.reason))
+        except urllib.error.HTTPError as e:
+            e.fp.read()
+            if e.status == 412:
+                # precondition failed -> lost race with other upload
+                raise HttpAlreadyExistsError()
+            raise HttpUploadError("PUT {} {}".format(e.status, e.reason))
 
     def _mkdir(self, path):
         # MKCOL resources must have a trailing slash because they are
         # directories. Otherwise Apache might send a HTTP 301. Nginx refuses to
         # create the directory with a 409 which looks odd.
         if not path.endswith("/"):
             path += "/"
-        connection = self._getConnection()
-        connection.request("MKCOL", path, headers=self._getHeaders())
-        response = connection.getresponse()
-        response.read()
-        return response
+
+        req = urllib.request.Request (self._getURL(path),
+                                      headers=self._getHeaders(), method="MKCOL")
+        try:
+            with urllib.request.urlopen (req, context=self.__context) as resp:
+                return (resp.status, None)
+        except urllib.error.HTTPError as e:
+            e.fp.read()
+            return (e.status, e.reason)
 
     def mkdir(self, path, depth=1):
         if depth > 0:
-            response = self._mkdir(path)
-            if response.status == 409:
+            status, reason = self._mkdir(path)
+            if status == 409:
                 (_path, _, _) = path.rpartition("/")
                 self.mkdir(_path, depth - 1)
-                response = self._mkdir(path)
+                status,reason = self._mkdir(path)
             # We expect to create the directory (201) or it already existed (405).
             # If the server does not support MKCOL we'd expect a 405 too and hope
             # for the best...
-            if response.status not in [201, 405]:
-                raise HttpUploadError("MKCOL {} {}".format(response.status, response.reason))
+            if status not in [201, 405]:
+                raise HttpUploadError("MKCOL {} {}".format(status, reason))
 
     def listdir(self, path):
         base_path = self.__url.path
-        # create a full path ending with trailing / (should prevent http 301 - moved permanently)
-        path = '/'.join([base_path, path.strip('/'), ''])
```

Comment:
Just to be sure: can we really remove these lines? I think they are crucial...

Reply:
Looks like you are right. I did some testing against jfrog-artifactory this morning resulting in a backtrace with `bob.webdav.HttpDownloadError: PROPFIND 302 Moved Temporarily`. I can't remember why I removed this :thinking:  I restored these lines and added catches for the HTTPException in stat and listdir.

There was also a problem in  `_getURL` producing invalid paths for artifactory, e.g. `http://foo/archive//.` instead of `http://foo/archive/.` causing a 404 -> empty listdir -> backtrace. I added fixes for both issues.